File: apps/ios/Shared/Views/Badges/SupportSimpleXBanner.swift
//
//  SupportSimpleXBanner.swift
//  SimpleX (iOS)
//
//  Created by spaced4ndy on 28.07.2026.
//  Copyright © 2026 SimpleX Chat. All rights reserved.
//

import SwiftUI
import SimpleXChat

struct SupportSimpleXBanner: View {
    @EnvironmentObject var theme: AppTheme
    @Environment(\.colorScheme) var colorScheme: ColorScheme
    var title: LocalizedStringKey = "Support SimpleX"
    var subtitle: LocalizedStringKey = "Get badge + better files"
    let onTap: () -> Void
    let onDismiss: () -> Void

    // the card's own height, for centring the fallback hero; hero stays fixed so its above-card
    // overhang shrinks at very large fonts
    @ScaledMetric(relativeTo: .body) private var scaledCardHeight: CGFloat = 72
    private var cardHeight: CGFloat { max(72, scaledCardHeight) }
    private let heroWidth: CGFloat = 110
    // shorter than the natural drawn height so .clipped() slices the phone body at card bottom
    private let heroVisibleHeight: CGFloat = 108
    // hero right edge sits exactly at the dismiss X's left edge (X trailing 16 + width 12)
    private let heroTrailingPadding: CGFloat = 28
    private let textToHeroGap: CGFloat = 6

    var body: some View {
        // Card is the Button; hero is an overlay so it can extend above the card top without
        // affecting layout size. Dismiss X is a ZStack sibling anchored to the card's top-right.
        ZStack(alignment: .topTrailing) {
            Button(action: onTap) {
                HStack(spacing: 0) {
                    VStack(alignment: .leading, spacing: 4) {
                        Text(title)
                            .font(.headline)
                            .foregroundColor(theme.colors.primary)
                            .lineLimit(2)
                        Text(subtitle)
                            .font(.subheadline)
                            .foregroundColor(theme.colors.onBackground)
                            .lineLimit(2)
                    }
                    Spacer(minLength: heroWidth + heroTrailingPadding + textToHeroGap)
                }
                .modifier(BannerCard())
            }
            .buttonStyle(.plain)
            .overlay(alignment: .bottomTrailing) {
                heroThumbnail()
                    .padding(.trailing, heroTrailingPadding)
                    .allowsHitTesting(false)
            }

            BannerDismissButton(onDismiss: onDismiss)
        }
    }

    @ViewBuilder
    private func heroThumbnail() -> some View {
        #if SIMPLEX_ASSETS
        // draws at natural aspect, top-aligned in a shorter slot; .clipped() cuts the overflow at card bottom
        Image(colorScheme == .light ? "phone-supporter" : "phone-supporter-light")
            .resizable()
            .aspectRatio(contentMode: .fill)
            .frame(width: heroWidth, height: heroVisibleHeight, alignment: .top)
            .clipped()
        #else
        Image("badge-supporter")
            .resizable()
            .scaledToFit()
            .frame(width: 48, height: 48)
            .padding(.vertical, (cardHeight - 48) / 2)
            .padding(.trailing, 12)
        #endif
    }
}

struct SupportSimpleXBanner_Previews: PreviewProvider {
    static var previews: some View {
        SupportSimpleXBanner(onTap: {}, onDismiss: {})
            .padding()
    }
}

File: apps/multiplatform/common/src/commonMain/kotlin/chat/simplex/common/views/badges/SupportSimpleXBanner.kt
package chat.simplex.common.views.badges

import androidx.compose.foundation.*
import androidx.compose.foundation.layout.*
import androidx.compose.material.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.layout.Layout
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import dev.icerock.moko.resources.compose.painterResource
import chat.simplex.common.BuildConfigCommon
import chat.simplex.common.platform.*
import chat.simplex.common.ui.theme.*
import chat.simplex.common.views.chatlist.BannerDismissButton
import chat.simplex.common.views.chatlist.bannerCard
import chat.simplex.common.views.helpers.*
import chat.simplex.res.MR

@Composable
fun SupportSimpleXBanner(
  title: String = generalGetString(MR.strings.badges_banner_title),
  subtitle: String = generalGetString(MR.strings.badges_banner_subtitle),
  onTap: () -> Unit,
  onDismiss: () -> Unit
) {
  // the card's own height, for centring the fallback hero; hero stays fixed so its above-card
  // overhang shrinks at very large fonts
  val cardHeight = (72.dp * fontSizeMultiplier).coerceAtLeast(72.dp)
  val cardTrailingPadding = 8.dp
  val heroWidth = 110.dp
  // shorter than the natural drawn height so ContentScale.Crop slices the phone body at card bottom
  val heroVisibleHeight = 108.dp
  // hero right edge sits at the dismiss X's icon left edge (X: outer 4pt + inner-pad 8 + half of 16pt icon)
  val heroTrailingPadding = 28.dp
  val textToHeroGap = 6.dp

  // Layout sizes to the card; hero is placed at y = cardHeight - heroHeight (negative → hero
  // overhangs above card at normal fonts, 0/positive → hero fits inside card at large fonts).
  Layout(content = {
    Box(Modifier.fillMaxWidth()) {
      Row(
        Modifier
          .bannerCard(onTap)
          .padding(
            start = 16.dp,
            end = cardTrailingPadding + heroWidth + heroTrailingPadding + textToHeroGap,
            top = 12.dp,
            bottom = 12.dp
          ),
        verticalAlignment = Alignment.CenterVertically
      ) {
        Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
          Text(
            title,
            style = MaterialTheme.typography.body1,
            fontWeight = FontWeight.SemiBold,
            color = MaterialTheme.colors.primary,
            maxLines = 2,
            overflow = TextOverflow.Ellipsis
          )
          Text(
            subtitle,
            style = MaterialTheme.typography.body2,
            color = MaterialTheme.colors.onBackground,
            maxLines = 2,
            overflow = TextOverflow.Ellipsis
          )
        }
      }

      BannerDismissButton(Modifier.align(Alignment.TopEnd), onDismiss)
    }

    HeroThumbnail(
      heroWidth = heroWidth,
      heroVisibleHeight = heroVisibleHeight,
      cardHeight = cardHeight,
      trailingPadding = heroTrailingPadding
    )
  }) { measurables, constraints ->
    val cardPlaceable = measurables[0].measure(constraints)
    val heroPlaceable = measurables[1].measure(constraints.copy(minWidth = 0, minHeight = 0))
    layout(cardPlaceable.width, cardPlaceable.height) {
      cardPlaceable.place(0, 0)
      heroPlaceable.place(cardPlaceable.width - heroPlaceable.width, cardPlaceable.height - heroPlaceable.height)
    }
  }
}

@Composable
private fun HeroThumbnail(heroWidth: Dp, heroVisibleHeight: Dp, cardHeight: Dp, trailingPadding: Dp) {
  if (BuildConfigCommon.SIMPLEX_ASSETS) {
    // draws at natural aspect, top-aligned in a shorter slot; ContentScale.Crop cuts the overflow at card bottom
    Image(
      painterResource(if (isInDarkTheme()) MR.images.phone_supporter_light else MR.images.phone_supporter),
      contentDescription = null,
      contentScale = ContentScale.Crop,
      alignment = Alignment.TopCenter,
      modifier = Modifier.padding(end = trailingPadding).size(width = heroWidth, height = heroVisibleHeight)
    )
  } else {
    val badgeSize = 48.dp
    Image(
      painterResource(MR.images.badge_supporter),
      contentDescription = null,
      contentScale = ContentScale.Fit,
      modifier = Modifier
        .padding(end = trailingPadding + 12.dp, top = (cardHeight - badgeSize) / 2, bottom = (cardHeight - badgeSize) / 2)
        .size(badgeSize)
    )
  }
}
